pkg/cluster/cluster.go

```diff
@@ -205,6 +209,45 @@ func (p *Peer) Join(peerType PeerType, initialMetadata PeerMetadata) error {
 		Metadata:     initialMetadata,
 	})
 
+	if p.refreshInterval != 0 {
+		go p.periodicallyRefresh()
+	}
+
+	return nil
+}
+
+func (p *Peer) periodicallyRefresh() {
+	tick := time.NewTicker(p.refreshInterval)
+	defer tick.Stop()
+
+	for {
+		select {
+		case <-p.stopc:
+			return
+		case <-tick.C:
+			if err := p.Refresh(); err != nil {
+				level.Error(p.logger).Log("msg", "Refreshing memberlist", "err", err)
+			}
+		}
+	}
+}
+
+// Refresh renews membership cluster, this will refresh DNS names and join newly added members
+func (p *Peer) Refresh() error {
+	p.mlistMtx.Lock()
+	defer p.mlistMtx.Unlock()
+
+	if p.mlist == nil {
+		return nil
+	}
+
+	prev := p.mlist.NumMembers()
+	curr, err := p.mlist.Join(p.knownPeers)
```

Comment:
It is not trivial here -> basically you join member that already joined. I need to double check the consequences, but I am kind of worried that:
- peer is not closed until you create new one (join)
- even if we met above it will meet a little bit distruption.

have you looked closely what is the beaviour of this for memberlist?

Reply:
@Bplotka I've looked thru code and it won't close the peer. But you are right about the disruption, as it calls pushPullNode (https://github.com/hashicorp/memberlist/blob/master/memberlist.go#L225), which is rather expensive operation, as per docs:
```
pushPull is invoked periodically to randomly perform a complete state
exchange. Used to ensure a high level of convergence, but is also
 reasonably expensive as the entire state of this node is exchanged
 with the other node.


```

 I'll change that we just join new members, just by resolving dns and looping thru current members